File: ios/ReactNativeSignView/SignView.m
//
//  SignView.m
//  ReactNativeSignView
//
//  Created by Harish on 20/02/19.
//  Copyright © 2019 Tekion. All rights reserved.
//

#import "SignView.h"
#include <math.h>

@implementation SignView{
    UIBezierPath* currentPath;
    CAShapeLayer* pathLayer;
}
@synthesize delegate;

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self commonInit];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        [self commonInit];
    }
    return self;
}

- (void) commonInit{
    self.backgroundColor = nil;
    pathLayer = [[CAShapeLayer alloc] init];
    currentPath = [[UIBezierPath alloc] init];
    pathLayer.path = [currentPath CGPath];
    pathLayer.strokeColor = [UIColor blackColor].CGColor;
    pathLayer.lineWidth = 2;
    [self.layer addSublayer:pathLayer];
    
    UIGestureRecognizer* tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(onTap:)];
    [self addGestureRecognizer:tap];
}

- (void)onTap:(UITapGestureRecognizer*)sender {
    CGPoint location = [sender locationInView:self];
    [currentPath moveToPoint:location];
    [currentPath addArcWithCenter:location radius:1 startAngle:0 endAngle:(M_PI * 2) clockwise:true];
    [self drawPaths];
    [self onSignatureChange];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event{
    UITouch *touch = [[event allTouches] anyObject];
    CGPoint location = [touch locationInView:touch.view];
    [currentPath moveToPoint:location];
}

- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event{
    UITouch *touch = [[event allTouches] anyObject];
    CGPoint location = [touch locationInView:touch.view];
    
    CGFloat width = [self bounds].size.width;
    CGFloat height = [self bounds].size.height;
    
    if(location.x > 0 && location.x < width &&
       location.y > 0 && location.y < height){
        [currentPath addLineToPoint:location];
        [currentPath moveToPoint:location];
        [self drawPaths];
    }
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event{
    UITouch *touch = [[event allTouches] anyObject];
    CGPoint location = [touch locationInView:touch.view];
    [currentPath moveToPoint:location];
    [self drawPaths];
    [self onSignatureChange];
}

-(void)drawPaths{
    pathLayer.path = [currentPath CGPath];
}

-(UIImage*) rotateImage:(UIImage *)image : (CGFloat) rotation{
    CGAffineTransform t = CGAffineTransformMakeRotation(rotation);
    CGRect sizeRect = (CGRect) {.size = image.size};
    CGRect destRect = CGRectApplyAffineTransform(sizeRect, t);
    CGSize destinationSize = destRect.size;
    
    UIGraphicsBeginImageContext(destinationSize);
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextTranslateCTM(context, destinationSize.width / 2.0f, destinationSize.height / 2.0f);
    CGContextRotateCTM(context, rotation);
    [image drawInRect:CGRectMake(-image.size.width / 2.0f, -image.size.height / 2.0f, image.size.width, image.size.height)];
    
    UIImage *newImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return newImage;
}


-(UIImage *)getSignatureImage{
    if(currentPath.isEmpty) return nil;
    else{
        CGRect signBounds = currentPath.bounds;
        CGFloat width = signBounds.origin.x + signBounds.size.width;
        CGFloat height = signBounds.origin.y + signBounds.size.height;
        
        UIGraphicsBeginImageContext(CGSizeMake(width, height));
        [pathLayer renderInContext:UIGraphicsGetCurrentContext()];
        UIImage* signatureImage = UIGraphicsGetImageFromCurrentImageContext();
        UIGraphicsEndImageContext();
        
        CGRect signatureBoundingRect = [currentPath bounds];
        CGImageRef imageRef = CGImageCreateWithImageInRect(signatureImage.CGImage, signatureBoundingRect);
        UIImage *croppedImage   = [UIImage imageWithCGImage:imageRef];
        CGImageRelease(imageRef);
        
        return croppedImage;
    }
}

-(NSString *)getBase64StringOfImage:(UIImage *)image{
    NSData* nsDataOfImage = UIImagePNGRepresentation(image);
    NSString * base64StringOfImage = [nsDataOfImage base64EncodedStringWithOptions:0];
    return base64StringOfImage;
}

-(void)onSignatureChange{
    if([self delegate] != nil){
        UIImage* image = [self getSignatureImage];
        NSString * base64StringOfImage = [self getBase64StringOfImage:image];
        [self.delegate onSignatureChangesCb:base64StringOfImage];
    }
}

- (void)setSignatureColor: (UIColor*) signColor{
    pathLayer.strokeColor = signColor.CGColor;
}

- (void)clearSignature{
    [currentPath removeAllPoints];
    [self drawPaths];
}

@end

File: ios/ReactNativeSignView/SignView.swift
//
//  CircularView.swift
//  CustomViewSample
//
//  Created by Harish on 09/02/19.
//

// THis is native view for handling signature. Do not add react-native stuffs here.


import Foundation
import UIKit

class SignView: UIView {
    
    public var delegate: SignViewCallbacks?;
    
    private var currentPath = UIBezierPath();
    private let pathLayer = CAShapeLayer();
    
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        commonInit()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder);
        commonInit();
    }
    
    func commonInit(){
        self.backgroundColor = UIColor.white;
        pathLayer.path = currentPath.cgPath;
        pathLayer.strokeColor = UIColor.black.cgColor;
        pathLayer.lineWidth = 2;
        self.layer.addSublayer(pathLayer);
        
        let tap = UITapGestureRecognizer(target: self, action: #selector(onTap))
        self.addGestureRecognizer(tap)
    }
    
    @objc public func onTap(sender: UITapGestureRecognizer? = nil) {
        let location = sender!.location(in: self);
        currentPath.move(to: location);
        currentPath.addArc(withCenter: location, radius: 1, startAngle: 0, endAngle: .pi * 2, clockwise: true);
        drawPaths();
        self.onSignatureChange();
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        let touch = touches.first!
        let location = touch.location(in: self)
        currentPath.move(to: location);
    }
    
    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        let touch = touches.first!
        let location = touch.location(in: self)
        currentPath.move(to: location);
        drawPaths();
        self.onSignatureChange();
    }
    
    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        if let touch = touches.first {
            let location = touch.location(in: self);
            
            if(location.x > 0 && location.x < self.bounds.width &&
                location.y > 0 && location.y < self.bounds.height){
                self.currentPath.addLine(to: location);
                self.currentPath.move(to: location);
                drawPaths();
            }
        }
    }
    
    private func getSignatureImage() -> UIImage? {
        if(currentPath.isEmpty){
            return nil;
        } else{
            UIGraphicsBeginImageContext(CGSize(width: self.bounds.size.width, height: self.bounds.size.height))
            self.pathLayer.render(in: UIGraphicsGetCurrentContext()!)
            let signatureImage = UIGraphicsGetImageFromCurrentImageContext()!
            UIGraphicsEndImageContext();
            return signatureImage;
        }
    }
    
    private func getBase64StringOfImage(_ image:UIImage) -> String? {
        let nsDataOfImage = image.pngData();
        let base64StringOfImage = nsDataOfImage?.base64EncodedString();
        return base64StringOfImage;
    }
    
    private func onSignatureChange(){
        var base64StringOfImage = "";
        if let signatureImage = getSignatureImage(){
            base64StringOfImage =  getBase64StringOfImage(signatureImage) ?? "";
        }
        delegate?.onSignatureChangesCb(base64StringOfImage);
    }
    
    @objc public func clearSignature(){
        currentPath.removeAllPoints();
        drawPaths();
        self.onSignatureChange();
    }
    
    @objc public func setSignatureColor(_ signColor:UIColor){
        pathLayer.strokeColor = signColor.cgColor;
    }
    
    func drawPaths(){
        pathLayer.path = currentPath.cgPath;
    }
}



